int32 flipper
float32 time